Moved ownership of extractor_ and range_filter_ from RangeScanOptions…

erikleitch opened the pull request:
… to RangeScanTask to simplify memory management and fix the memory leak introduced by extractor_ not being deleted in ~RangeScanOption()

Commit:
Moved ownership of extractor_ and range_filter_ from RangeScanOptions to RangeScanTask to simplify memory management and fix the memory leak introduced by extractor_ not being deleted in ~RangeScanOption()

erikleitch commented:
This is an internal change to eleveldb (no repercussion for erlang/elevldb interface) to fix a small memory leak I've identified in the C++ layer.

erikleitch commented:
Some additional context: this was a problem that originated with the original quick-n-dirty TS prototype: in the protoype/timeseries-range-scan-filter branch, these pointers belonged to RangeScanOptions, and were allocated but never deleted anywhere.  Because RangeScanOptions gets passed to RangeScanTask and copied into that object,  an implicit copy constructor for RangeScanOptions was getting called that duplicated those pointers, making it unsafe to delete those members in ~RangeScanOptions().  The purpose of this PR is to:

1) move responsibility for allocation and deletion of these pointers to RangeScanTask, which is the object that actually uses them

2) by changing ownership, ensure that the pointers are allocated once and never copied

3) explicitly free these objects in ~RangeScanTask(), ensuring that the memory for these objects is released

matthewvon commented on c_src/workitems.cc:
```diff
@@ -487,19 +487,40 @@ RangeScanTask::RangeScanTask(ErlNifEnv * caller_env,
+        if(options_.encodingType_ == Encoding::MSGPACK)
+            extractor_ = new ExtractorMsgpack();
+        else {
+            ThrowRuntimeError("An invalid object encoding was specified");
```
streaming_start() in eleveldb.cc calls "new" operator on this object without try/catch.  Riak will crash if this throws.

matthewvon commented on c_src/workitems.h:
```diff
@@ -629,6 +619,8 @@ namespace eleveldb {
         std::string end_key_;
         bool has_end_key_;
         SyncObject * sync_obj_;
```
Appears sync_obj_ is NOT optional.  Therefore it needs to be a reference, not a pointer.  If optional, pointer is not validated for NULL and could cause segfault.

matthewvon commented on c_src/workitems.cc:
```diff
@@ -791,9 +812,9 @@ work_result RangeScanTask::operator()()
-                                                    options_.rangeFilterSpec_, *(options_.extractor_),
+                                                    options_.rangeFilterSpec_, *(extractor_),
                                                     throwIfBadFilter);
                     } catch(std::runtime_error& err) {
```
This is going to throw.  I believe it is throwing to threading.cc notify_caller() which is called by eleveldb_write_thread_worker().  I do not see try/catch block anywhere in that call stack.

No, there is a surrounding try/catch I missed.

Commit:
Added protection for potential throw in RangeScanTask constructor, added check for passed SyncObject, modified thread pool sizes

erikleitch commented:
Committed a few changes to feature/bugfix-extractorleak in response to Matthew's comments:

1) added protection around the RangeScanTask constructor, which can now in principle throw

2) added check for NULL SyncObj on construction (I don't want to rampage through that code further until I understand more about how that obj is used)

3) made threadpool sizes prime

matthewvon commented:
+1
